packages/ecs-agent/ecs-agent.spec

```diff
@@ -48,23 +76,92 @@ Requires: %{_cross_os}iptables
+#     │   └── [unpacked sources]
+#     └── GOPATH
+#         └── src/github.com/aws
+#             ├── amazon-ecs-agent [symlink]
+#             ├── amazon-ecs-cni-plugins [symlink]
+#             └── amazon-vpc-cni-plugins [symlink]
 
+# Extract Source0, which has a top-level directory of
+# %{agent_gorepo}-%{agent_gover}
+# -c: Create directory (%{name}-%{version})
+# -q: Unpack quietly
+%setup -c -q
+# Change to the directory that we unpacked
+cd %{agent_gorepo}-%{agent_gover}
+# Set up git so we can apply patches
+# This is included in autosetup, but not autopatch
+%global __scm git
+%__scm_setup_git
+# Apply patches up to 0999
+%autopatch -M 0999
 # Replace upstream's version.go to support build-time values from ldflags. This
 # avoids maintenance of patches that use always changing version-control tokens
 # in its replacement.
-cp %{S:9} "agent/version/version.go"
+cp %{S:109} "agent/version/version.go"
+
+# Extract Source1, which has a top-level directory of
+# %{ecscni_gorepo}-%{ecscni_gitrev}
+# -T: Do not perform default archive unpack (i.e., skip Source0)
+# -D: Do not delete directory before unpacking sources (i.e., don't delete
+#     unpacked Source0)
+# -a: Unpack after changing into the directory
+# -q: Unpack quietly
+# See http://ftp.rpm.org/max-rpm/s1-rpm-inside-macros.html
+%setup -T -D -a 1 -q
+# Change to the directory that we unpacked
+cd %{ecscni_gorepo}-%{ecscni_gitrev}
+# Set up git so we can apply patches
+# This is included in autosetup, but not autopatch
+%__scm_setup_git
+# Apply patches from 1000 to 1999
+%autopatch -m 1000 -M 1999
+
+# Extract Source2, which has a top-level directory of
+# %{vpccni_gorepo}-%{vpccni_gitrev}
+# -T: Do not perform default archive unpack (i.e., skip Source0)
+# -D: Do not delete directory before unpacking sources (i.e., don't delete
+#     unpacked Source0)
+# -a: Unpack after changing into the directory
+# -q: Unpack quietly
+# See http://ftp.rpm.org/max-rpm/s1-rpm-inside-macros.html
+%setup -T -D -a 2 -q
+# Change to the directory that we unpacked
+cd %{vpccni_gorepo}-%{vpccni_gitrev}
+# Set up git so we can apply patches
+# This is included in autosetup, but not autopatch
+%__scm_setup_git
+# Apply patches from 1000 to 1999
+%autopatch -m 2000 -M 2999
+
+cd ../
+# Symlink amazon-ecs-agent-%{agent_gover} to the GOPATH location
+%cross_go_setup %{name}-%{version}/%{agent_gorepo}-%{agent_gover} %{agent_goproject} %{agent_goimport}
+# Symlink amazon-ecs-cni-plugins-%{ecscni_gitrev} to the GOPATH location
+%cross_go_setup %{name}-%{version}/%{ecscni_gorepo}-%{ecscni_gitrev} %{ecscni_goproject} %{ecscni_goimport}
+# Symlink amazon-vpc-cni-plugins-%{vpccni_gitrev} to the GOPATH location
+%cross_go_setup %{name}-%{version}/%{vpccni_gorepo}-%{vpccni_gitrev} %{vpccni_goproject} %{vpccni_goimport}
 
 %build
+pushd $(pwd)
 # Build the agent
-%cross_go_configure %{goimport}
+# cross_go_configure cd's to the correct GOPATH location
+%cross_go_configure %{agent_goimport}
```

Comment:
Are the `pushd` and `popd` working around a limitation in `%cross_go_configure` where it can't `cd` to the right path *unless* it is at the top of the unpacked tree?

I'd prefer to capture that path and later `cd` *directly* to the path. This avoids later edits that would need to consider how they might affect working directory navigation.

```specfile
%build
BUILD_TOP=$(pwd -P)

# build a
cd $BUILD_TOP
%cross_go_configure %{a_goimport}

# build a/child/artifact
cd child/artifact
%make

# build b
cd $BUILD_TOP
%cross_go_configure %{b_goimport}

...

cd $BUILD_TOP
```

Reply:
> Are the `pushd` and `popd` working around a limitation in `%cross_go_configure` where it can't `cd` to the right path unless it is at the _top_ of the unpacked tree?

Yes

> I'd prefer to capture that path and later `cd` _directly_ to the path. This avoids later edits that would need to consider how they might affect working directory navigation.

I like that suggestion.  I'll make sure it works and then update here.

Reply:
Done